### ISSUE TITLE
SGE Fails to Process Gift After NPSP 3.187 Release

### ISSUE THREAD

**Issue by jesslopez-sf** [SalesforceFoundation/gem#285]
**Note: FOR INTERNAL USE ONLY. For support questions, please use the GEM Chatter group in [The Power of Us Hub](https://powerofus.force.com).
For feature requests, please use the Ideas tab in [The Power of Us Hub](https://powerofus.force.com).**

* **Do you want to request a *feature* or report a *bug*?**
Bug

* **What is the current behavior?**
When processing a gift via Single Gift Entry, the following error appears: "SObject row was retrieved via SOQL without querying the requested field: npsp__DataImport__c.npsp__NPSP_Data_Import_Batch__r". The gift is not created.

* **What is the expected behavior?**
The gift is processed without any errors.

* **What is the motivation / use case for changing the behavior?**
Single Gift Entry should work regardless of npsp updates.

* **Acceptance Criteria**

  - **Given**: I am on the Single Gift Entry page
  - **When**: I enter information into the form and click the Save button
  - **Then**: The Opportunity and any other information entered into the form should be saved without error.

### Root Cause:
Debug Logs show that this is occurring in the npsp shouldSetDonationBatchNumber() field within the BDI_Donations class. This method checks to see if we’re in dry run mode, if the NPSP_Data_Import_Batch__c field is blank or if the NPSP Data Import Batch’s Batch_Number__c field is blank. 

The issue occurs because gem does not send the Batch’s Batch_Number__c field to npsp BDI to process. The batch is also blank.

The fields we set in GEM are in the runGiftProcess() method of the GiftEntryFormController class. And this is the query that is returned in the log:
```
12:48:48.0 (905403421)|SOQL_EXECUTE_BEGIN|[737]|Aggregations:0|SELECT Id,OwnerId,IsDeleted,Name,CreatedDate,CreatedById,LastModifiedDate,LastModifiedById,SystemModstamp,LastViewedDate,LastReferencedDate,npsp__Account1ImportStatus__c,npsp__Account1Imported__c,npsp__Account1_City__c,npsp__Account1_Country__c,npsp__Account1_Name__c,npsp__Account1_Phone__c,npsp__Account1_State_Province__c,npsp__Account1_Street__c,npsp__Account1_Website__c,npsp__Account1_Zip_Postal_Code__c,npsp__Account2ImportStatus__c,npsp__Account2Imported__c,npsp__Account2_City__c,npsp__Account2_Country__c,npsp__Account2_Name__c,npsp__Account2_Phone__c,npsp__Account2_State_Province__c,npsp__Account2_Street__c,npsp__Account2_Website__c,npsp__Account2_Zip_Postal_Code__c,npsp__Additional_Object_JSON__c,npsp__ApexJobId__c,npsp__Campaign_Member_Status__c,npsp__Contact1ImportStatus__c,npsp__Contact1Imported__c,npsp__Contact1_Alternate_Email__c,npsp__Contact1_Birthdate__c,npsp__Contact1_Consent_Message__c,npsp__Contact1_Consent_Opt_In__c,npsp__Contact1_Firstname__c,npsp__Contact1_Home_Phone__c,npsp__Contact1_Lastname__c,npsp__Contact1_Mobile_Phone__c,npsp__Contact1_Other_Phone__c,npsp__Contact1_Personal_Email__c,npsp__Contact1_Preferred_Email__c,npsp__Contact1_Preferred_Phone__c,npsp__Contact1_Salutation__c,npsp__Contact1_Title__c,npsp__Contact1_Work_Email__c,npsp__Contact1_Work_Phone__c,npsp__Contact2ImportStatus__c,npsp__Contact2Imported__c,npsp__Contact2_Alternate_Email__c,npsp__Contact2_Birthdate__c,npsp__Contact2_Firstname__c,npsp__Contact2_Home_Phone__c,npsp__Contact2_Lastname__c,npsp__Contact2_Mobile_Phone__c,npsp__Contact2_Other_Phone__c,npsp__Contact2_Personal_Email__c,npsp__Contact2_Preferred_Email__c,npsp__Contact2_Preferred_Phone__c,npsp__Contact2_Salutation__c,npsp__Contact2_Title__c,npsp__Contact2_Work_Email__c,npsp__Contact2_Work_Phone__c,npsp__DonationCampaignImportStatus__c,npsp__DonationCampaignImported__c,npsp__DonationImportStatus__c,npsp__DonationImported__c,npsp__Donation_Amount__c,npsp__Donation_Campaign_Name__c,npsp__Donation_Date__c,npsp__Donation_Description__c,npsp__Donation_Donor__c,npsp__Donation_Elevate_Recurring_ID__c,npsp__Donation_Member_Level__c,npsp__Donation_Membership_End_Date__c,npsp__Donation_Membership_Origin__c,npsp__Donation_Membership_Start_Date__c,npsp__Donation_Name__c,npsp__Donation_Possible_Matches__c,npsp__Donation_Record_Type_Name__c,npsp__Donation_Stage__c,npsp__Donation_Type__c,npsp__Elevate_Custom_Metadata__c,npsp__FailureInformation__c,npsp__GAU_Allocation_1_Amount__c,npsp__GAU_Allocation_1_GAU__c,npsp__GAU_Allocation_1_Import_Status__c,npsp__GAU_Allocation_1_Imported__c,npsp__GAU_Allocation_1_Percent__c,npsp__GAU_Allocation_2_Amount__c,npsp__GAU_Allocation_2_GAU__c,npsp__GAU_Allocation_2_Import_Status__c,npsp__GAU_Allocation_2_Imported__c,npsp__GAU_Allocation_2_Percent__c,npsp__HomeAddressImportStatus__c,npsp__HomeAddressImported__c,npsp__Home_City__c,npsp__Home_Country__c,npsp__Home_State_Province__c,npsp__Home_Street__c,npsp__Home_Zip_Postal_Code__c,npsp__HouseholdAccountImported__c,npsp__Household_Phone__c,npsp__ImportedDate__c,npsp__NPSP_Data_Import_Batch__c,npsp__Opportunity_Contact_Role_1_ImportStatus__c,npsp__Opportunity_Contact_Role_1_Imported__c,npsp__Opportunity_Contact_Role_1_Role__c,npsp__Opportunity_Contact_Role_2_ImportStatus__c,npsp__Opportunity_Contact_Role_2_Imported__c,npsp__Opportunity_Contact_Role_2_Role__c,npsp__PaymentImportStatus__c,npsp__PaymentImported__c,npsp__Payment_Authorization_Token__c,npsp__Payment_Authorized_Date__c,npsp__Payment_Authorized_UTC_Timestamp__c,npsp__Payment_Card_Expiration_Month__c,npsp__Payment_Card_Expiration_Year__c,npsp__Payment_Card_Last_4__c,npsp__Payment_Card_Network__c,npsp__Payment_Check_Reference_Number__c,npsp__Payment_Declined_Reason__c,npsp__Payment_Donor_Cover_Amount__c,npsp__Payment_Elevate_Created_Date__c,npsp__Payment_Elevate_Created_UTC_Timestamp__c,npsp__Payment_Elevate_ID__c,npsp__Payment_Elevate_Original_Payment_ID__c,npsp__Payment_Gateway_ID__c,npsp__Payment_Gateway_Payment_ID__c,npsp__Payment_Method__c,npsp__Payment_Origin_ID__c,npsp__Payment_Origin_Name__c,npsp__Payment_Origin_Type__c,npsp__Payment_Possible_Matches__c,npsp__Payment_Status__c,npsp__Payment_Type__c,npsp__Status__c,gem__Do_Not_Automatically_Create_Payment__c,gem__Donation_Acknowledgment_Status__c,gem__Donation_Honoree_Contact__c,gem__Donation_Honoree_Name__c,gem__Donation_Matching_Gift_Account__c,gem__Donation_Matching_Gift_Employer__c,gem__Donation_Matching_Gift_Status__c,gem__Donation_Matching_Gift__c,gem__Donation_Notification_Message__c,gem__Donation_Notification_Recipient_Contact__c,gem__Donation_Tribute_Type__c,gem__Notification_Recipient_Name__c,gem__Post_Process_Object_JSON__c FROM npsp__DataImport__c WHERE Id = :diObjId
```

### Steps to Repeat
1. Spin up gem scratch org.
2. Create a Contact
3. Open the Single Gift Entry tab
4. Enter a gift for the Contact and click Save.

### Potential Solution
In npsp's shouldSetDonationBatchNumber() method, only check to see if the Batch's Batch_Number__c field is blank IF the batch field is not blank.

**Comment by jesslopez-sf** [comment]
W-8351099

**Comment by jesslopez-sf** [comment]
Included in NPSP 3.188 release
